### ISSUE TITLE
Maintain Version Control for same Project Name

### ISSUE THREAD

**Issue by priyankachaturvedi30** [Visual-Regression-Tracker/Visual-Regression-Tracker#244]
@pashidlos : Can there be a possibility through which we can maintain version control for our projects?
Say I have a Main Project P1 - 
I created one branch V1 and set my baselines on it, once my release V1 is out I would need to verify branch V2 (which may have some changes on top of V1), so in this particular case, I would want to clone data/baselines I created in V1 into my new branch V2 and then validate changes on V2. That would in turn help me keep my project's branches clean (V1 branch would have changes corresponding to release V1) and further changes and validation corresponding to version V2 can be done on V2 branch. 

Providing an option to Clone data from one branch to other would be really helpful to maintain version control.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 thanks for your request
it's implemented in a bit different manner similar to git flow 
assuming that feature branch is taken from main and then merged into it with changes
`main` branch those baselines are used for every `feature` branch unless diff found 
when you merge code from `feature` branch to `main` one - new baseline is created
after feature branch is merged you could remove whole test variation for feature branch 

could you describe why you want to copy baselines into new branch?

**Comment by priyankachaturvedi30** [comment]
@pashidlos : So, like standard releases, we have parallel ongoing releases for different customers. Say my customer 1 is on release ver 1 and customer 2 is on release ver 2 of same product. for such a case I would not want to overwrite baselines for any of the my customers. Thus I would like to copy baselines onto a new branch to keep the previous release baseline intact for customer that are still on previous versions

**Comment by pashidlos** [comment]
@priyankachaturvedi30 it's possible to do using different projects 
each project requires one main branch to compare new feature branches agains it 

does every customer version lives in separate branch or repository?

**Comment by priyankachaturvedi30** [comment]
Even if we do it using different projects I would still be needing to create baselines. For instance takethe below example
Base Branch -> Release 0 -> Feature 1
Customer A -> Release 1 -> Feature 1,2,3
Customer B -> Release 2 -> Feature 1,2
Customer C -> Release 3 -> Feature 1,2,3,4
In above example Feature 1 and Feature 2 is common for all customer across releases. And considering these 2 features comprises of 80% of my test cases. So the 80% baseline is still common across all 3 releases here. 
Now if a new Customer D comes in , assuming it requires some additional feature 5 on top of Feature 1,2
Customer D (New) -> Feature 1,2 (+ new feature 5)
So, in above scenario, i would simply copy baselines from Customer B and and update changes on those for Feature 5 releated updates.

This would help me create clean branches maintaining version control by providing me flexibilty to move back on any version and execute my tests on top of those.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 it sounds to me like merge feature 
you could trigger comparison for any feature branch agains main one 
it'll create new test run and you could and accept new baseline 
so this is some kind of copy baseline from feature branch to main 
the only difference that you would need an option to choose destination branch

**Comment by pashidlos** [comment]
I updated existing merge functionality to select other branches 
this is for testing purpose only so better to use separate instance for testing not to have your existing data corrupted
let me know if it works as expected
* `ui:4.12.0.244`
* `api:4.12.0.244`
* `migration:4.12.0`

**Comment by priyankachaturvedi30** [comment]
Sure @pashidlos : Much Appreciated. I'll verify and get back to you 👍

**Comment by priyankachaturvedi30** [comment]
So @pashidlos : I was trying to verify this merge feature but could not understand how to verify it. Can you explain a bit on how it works.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 go to project variations 
select source branch and enter destination branch name, then click merge
it'll create new test run where after approving baselines they are gonna be saved as destination branch baselines

**Comment by priyankachaturvedi30** [comment]
I tried doing so, but what I see it is basically creating a new test run under same project -- 
![image](https://user-images.githubusercontent.com/50797731/112001379-ecb11f80-8b44-11eb-8dd3-bc68986363fc.png)
Is it how it is supposed to work ?
That defeats my purpose of isolating and keeping the repository clean. It would have helped if I was able to copy the entire thing into a new instance (project name+branch) altogether.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 it's creating baseline only if there is diff with main branch 

> Base Branch -> Release 0 -> Feature 1
Customer A -> Release 1 -> Feature 1,2,3
Customer B -> Release 2 -> Feature 1,2
Customer C -> Release 3 -> Feature 1,2,3,4
Now if a new Customer D comes in , assuming it requires some additional feature 5 on top of Feature 1,2
Customer D (New) -> Feature 1,2 (+ new feature 5)

this allows you to merge Customer B -> Customer D to copy Feature 2

**Comment by priyankachaturvedi30** [comment]
Ok..so I tried this @pashidlos . Let me share my observation around here.
![image](https://user-images.githubusercontent.com/50797731/112015867-36543700-8b52-11eb-9002-377b87fab79f.png)
So, consider the above screenshot - Here I created a new project with branch master.
Step 1: I did a few baselining and executed 2 runs (Run1 an Run2). 
Step 2: After that I did a merge for release1 branch which created new test run # 3 with tag release1. 
Step 3: Next I executed Run3(for master) and updated the baseline for master branch. 
Step 4: Post doing that changes if i try running test against release1 following should happen:
1. The changes of master should not be there in release1. my idea here was that since i did changes for master branch only( in Step 3) and not for release1 (the merged branch which was created from Step 2 - So the baseline should be present only uptil step 2 and not post that ) , However This is not the case :( the baseline changes of Step 3 are there for release1 as well, those should not apply to release1 until i recreate copy(merge) post Step 3
2. My assumption it that since I am copying the baselines from Step2 (Till that point there is no Run1 build for release1) . So If I do a Run1 for release 1 if should work and not overwrite master Run1
Let me know if I am missing anything.

**Comment by priyankachaturvedi30** [comment]
@pashidlos : Did you get a chance to look at the above findings?

**Comment by priyankachaturvedi30** [comment]
@pashidlos : I tried out this lately, my bad I was missing the branch in which I wanted to merge the changes to. My observations stands correct as:
Master - > A.png
branch1 -> A.png and B.png
1. If **master** has baseline (A.png) , and I run my tests(A.png) on any new **branch1**, baseline from master would be used to verify the branches. (A.png would be compared and passed)
2. If **master** do not have those baselines and I run my tests(B.png) on new **branch1**, I have to create baseline in branches. (B.png would be new)
3. Changes of new **branch1** would be present in **master**, only after I merge the new **branch1** in master.
I understand this was a test version you provided to me (4.12.0.244), are we having these changes in the latest release that came over ?

**Comment by pashidlos** [comment]
@priyankachaturvedi30 this was not included in new releases

**Comment by priyankachaturvedi30** [comment]
Thanks for the update @pashidlos, This looks promising to me and offers a wonderful mechanism through which we can maintain the version control across builds. Can you try including this feature in any of your upcoming releases?

**Comment by priyankachaturvedi30** [comment]
Also, @pashidlos : In addition to this, it occurred to me that the PR Build that got auto-generated, could there be a provision that it can be approved\rejected and then merged by some other user? Just checking the possibility of this enhancement.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 I'll try to address this issue in scope of this ticket https://github.com/Visual-Regression-Tracker/Visual-Regression-Tracker/issues/287
